splipy/io/grdecl.py

```diff
@@ -99,146 +118,144 @@ def __enter__(self):
+        # Create the geometry
+        ngx, ngy, ngz = ngeom
+        b1 = BSplineBasis(p, [0]*(p-1) + [i/ngx for i in range(ngx+1)] + [1]*(p-1))
+        b2 = BSplineBasis(p, [0]*(p-1) + [i/ngy for i in range(ngy+1)] + [1]*(p-1))
+        b3 = BSplineBasis(p, [0]*(p-1) + [i/ngz for i in range(ngz+1)] + [1]*(p-1))
 
+        nx, ny, nz = self.n
+        l2_fit = surface_factory.least_square_fit
+        u, v, w = np.linspace(0, 1, nx+1), np.linspace(0, 1, ny+1), np.linspace(0, 1, nz+1)
+
+        if method == 'full':
+            bottom = l2_fit(self.c0_vol[:,:, 0,:].squeeze(), [b1, b2], [u, v])
+            top    = l2_fit(self.c0_vol[:,:,-1,:].squeeze(), [b1, b2], [u, v])
+            left   = l2_fit(self.c0_vol[ 0,:,:,:].squeeze(), [b2, b3], [v, w])
+            right  = l2_fit(self.c0_vol[-1,:,:,:].squeeze(), [b2, b3], [v, w])
+            front  = l2_fit(self.c0_vol[:, 0,:,:].squeeze(), [b1, b3], [u, w])
+            back   = l2_fit(self.c0_vol[:,-1,:,:].squeeze(), [b1, b3], [u, w])
+            volume = volume_factory.edge_surfaces([left, right, front, back, bottom, top])
+
+        elif method == 'z':
+            bottom = l2_fit(self.c0_vol[:,:, 0,:].squeeze(), [b1, b2], [u, v])
+            top    = l2_fit(self.c0_vol[:,:,-1,:].squeeze(), [b1, b2], [u, v])
+            volume = volume_factory.edge_surfaces([bottom, top])
+            volume.set_order(p)
+            volume.refine(ngx - 1, direction='w')
+
+        # Point-to-cell mapping
+        # TODO: Optimize more
+        u = [np.linspace(0,1,n) for n in ntexture]
+        points = volume(*u).reshape(-1, 3)
+        cellids = np.zeros(points.shape[:-1], dtype=int)
+        cell = None
+        for ptid, point in enumerate(tqdm(points, desc='Inverse mapping')):
+            i, j, k = cell = self.raw.cell_at(point, guess=cell)
+            cellid = i*ny*nz + j*nz + k
+            cellids[ptid] = cellid
+
+        cellids = cellids.reshape(tuple(len(param) for param in u))
 
         all_textures = {}
         for name in self.attribute:
-            one_texture = np.zeros(ntexture)
-            u = [np.linspace(0,1,n) for n in ntexture]
-            x = volume(*u)
-            x = np.reshape(x, [np.prod(ntexture),3])
-            for I in range(np.prod(ntexture)):
-                print(x[I,:])
-                i,j,k = self.raw.cell_at(x[I,:])
-                J = i*self.n[1]*self.n[2] + j*self.n[2] + k
-                one_texture[I] = self.attribute[name][J]
-            all_textures[name] = one_texture
-        return volume, all_textures
+            data = self.attribute[name][cellids]
+
+            # TODO: This flattens the image if it happens to be 3D (or higher...)
+            # do we need a way to communicate the structure back to the caller?
+            data = data.reshape(-1, data.shape[-1])
 
+            # TODO: This normalizes the image,
+            # but we need a way to communicate the ranges back to the caller
```

Comment:
Again I think we should return the un-scaled version from the grdecl-reader. One way to encode the ranges (and have control over the dimension) would be a "IFEM"-writer which either could be in the `io` directory or simply a stand-alone program which starts off with `from splipy import *`. This would dump the `g2`-geometry file and in addition the `xinp`-file with all material-texture inputs already filled in.

Reply:
Seems reasonable.